runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -12080,3 +12103,267 @@ J9::Z::TreeEvaluator::tabortEvaluator(TR::Node * node, TR::CodeGenerator * cg)
+ * BRC  call_snippet_to_jitReadBarrier (may use trampoline)
+ *
+ * Label: jitReadBarrier return label
+ * // reload evacuated reference
+ * LG   R_obj, 0(R_addr)
+ *
+ * doneLabel: endICF
+ * \endverbatim
+ *
+ *
+ * Unresolved reference fields load sequence for -XnocompressedRefs:
+ * \verbatim
+ *
+ * Label:   startICF
+ * LAY  R_addr, Ref_field_MemRef
+ * LG   R_obj, 0(R_addr)
+ *
+ * // range check with implicit CS cycle check
+ * CLG  R_obj, EvacuateBase(R_vmthread)
+ * BRC  COND_BL, doneLabel
+ * CLG  R_obj, EvacuateTop(R_vmthread)
+ * BRC  COND_BH, doneLabel
+ *
+ * STG  R_addr, floatTemp1(R_vmthread)
+ * BRC  call_snippet_to_jitReadBarrier (may use trampoline)
+ *
+ * Label: jitReadBarrier return label
+ * // reload evacuated reference
+ * LG   R_obj, 0(R_addr)
+ *
+ * doneLabel: endICF
+ * \endverbatim
+ *
+ * If compressed pointer is enabled, the LG instructions above are replaced by LLGF+SLLG.
+ */
+TR::Register *
+J9::Z::TreeEvaluator::generateSoftwareReadBarrier(TR::Node* node,
+                                                  TR::CodeGenerator* cg,
+                                                  TR::Register* resultReg,
+                                                  TR::MemoryReference* loadMemRef,
+                                                  const bool hasReturnReg)
+   {
+   TR::Compilation* comp = cg->comp();
+   TR::Register* locationReg = cg->allocateRegister();
+   TR::RealRegister* raReg   = cg->machine()->getRealRegister(cg->getReturnAddressRegister());
+   bool isCompressedRef = comp->useCompressedPointers();
+
+   if (!isCompressedRef)
+      {
+      TR::TreeEvaluator::checkAndSetMemRefDataSnippetRelocationType(node, cg, loadMemRef);
+      }
+
+   const bool fieldUnresolved = node->getSymbolReference()->isUnresolved();
+   if (comp->getOption(TR_TraceCG))
+      {
+      traceMsg(comp, "SoftwareReadBarrier: symbol is %s. Compr shift %d. RA reg: %s Entry reg %s\n",
+               fieldUnresolved ? "unresolved" : "resolved",
+               TR::Compiler->om.compressedReferenceShift(),
+               raReg->getRegisterName(comp),
+               cg->getEntryPointRealRegister()->getRegisterName(comp));
+      }
+
+   TR::LabelSymbol* startICFLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol* doneLabel = generateLabelSymbol(cg);
+
+   /*
+    * S390HelperCallSnippet clobbers return address register and may clobber entry point register if a trampoline
+    * is used. Make sure they are both free.
+    */
+   TR::Register* dummyRegForRA = cg->allocateRegister();
+   TR::Register* dummyRegForEntry = cg->allocateRegister();
+   dummyRegForRA->setPlaceholderReg();
+   dummyRegForEntry->setPlaceholderReg();
+
+   TR::RegisterDependencyConditions* deps = generateRegisterDependencyConditions(0, 7, cg);
+   deps->addPostCondition(resultReg, TR::RealRegister::AssignAny);
+   deps->addPostCondition(locationReg, TR::RealRegister::AssignAny);
+   deps->addPostCondition(dummyRegForRA, cg->getReturnAddressRegister());
```

Comment:
Now that `S390HelperCallSnippet` is no longer needed. I'll remove RA and rEP dependencies later.

Reply:
Removed `rEP` from dependencies in cfca0d4266.

Reply:
This thread is no longer valid. I switched back to `S390HelperCallSnippet`